File: src/com/quotegraph/model/vector/Vector2I.java
package com.quotegraph.model.vector;

import com.quotegraph.interfaces.IVector2I;

/**
 * 2D int vector implementation.
 *
 * @author Daniel Peters
 * @version 1.0
 */
public class Vector2I implements IVector2I {
  private int x;
  private int y;

  public Vector2I(int x, int y) {
    this.x = x;
    this.y = y;
  }

  @Override
  public int getX() {
    return this.x;
  }

  @Override
  public int getY() {
    return this.y;
  }

  @Override
  public void add(IVector2I vector) {
    this.x += vector.getX();
    this.y += vector.getY();
  }

  @Override
  public void sub(IVector2I vector) {
    this.x -= vector.getX();
    this.y -= vector.getY();
  }

  @Override
  public void mult(int scalar) {
    this.x *= scalar;
    this.y *= scalar;
  }

  @Override
  public void div(int scalar) {
    this.x /= scalar;
    this.y /= scalar;
  }

  @Override
  public int mag() {
    return (int) Math.sqrt(this.x * this.x + this.y * this.y);
  }

  @Override
  public void normalize() {

    int magnitude = mag();

    if (magnitude != 0) {
      div(magnitude);
    }
  }

  @Override
  public void limit(int max) {

    if (this.mag() > max) {

      this.normalize();
      this.mult(max);
    }
  }

  @Override
  public void set(int x, int y) {
    this.x = x;
    this.y = y;
  }

  @Override
  public void set(IVector2I vector) {
    this.x = vector.getX();
    this.y = vector.getY();
  }
}

File: src/main/kotlin/ch/peters/daniel/quotegraph/model/vector/Vector2I.kt
package ch.peters.daniel.quotegraph.model.vector

import ch.peters.daniel.quotegraph.interfaces.IVector2I

/**
 * 2D int vector implementation.
 *
 * @author Daniel Peters
 * @version 1.0
 */
class Vector2I(override var x: Int, override var y: Int) : IVector2I {
  override fun add(vector: IVector2I) {
    x += vector.x
    y += vector.x
  }

  override fun sub(vector: IVector2I) {
    x -= vector.x
    y -= vector.x
  }

  override fun mult(scalar: Int) {
    x *= scalar
    y *= scalar
  }

  override fun div(scalar: Int) {
    x /= scalar
    y /= scalar
  }

  override fun mag(): Int {
    return Math.sqrt((x * x + y * y).toDouble()).toInt()
  }

  override fun normalize() {
    val magnitude = mag()

    if (magnitude != 0) {
      div(magnitude)
    }
  }

  override fun limit(max: Int) {
    if (mag() > max) {
      normalize()
      mult(max)
    }
  }

  override fun set(valX: Int, valY: Int) {
    x = valX
    y = valY
  }

  override fun set(vector: IVector2I) {
    x = vector.x
    y = vector.x
  }
}
